lib/Extension/LanguageServerEvaluatableExpression/Handler/EvaluatableExpressionHandler.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerEvaluatableExpression\Handler;
+
+use function Amp\call;
+use Amp\Promise;
+use Microsoft\PhpParser\Parser;
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Token;
+use Phpactor\Extension\LanguageServerBridge\Converter\PositionConverter;
+use Phpactor\Extension\LanguageServerBridge\Converter\RangeConverter;
+use Phpactor\Extension\LanguageServerEvaluatableExpression\Protocol\EvaluatableExpression;
+use Phpactor\LanguageServerProtocol\Position;
+use Phpactor\LanguageServerProtocol\ServerCapabilities;
+use Phpactor\LanguageServerProtocol\TextDocumentIdentifier;
+use Phpactor\LanguageServer\Core\Handler\CanRegisterCapabilities;
+use Phpactor\LanguageServer\Core\Handler\Handler;
+use Phpactor\LanguageServer\Core\Workspace\Workspace;
+use Phpactor\TextDocument\TextDocumentBuilder;
+use Phpactor\WorseReflection\Core\Util\NodeUtil;
+use Phpactor\TextDocument\ByteOffsetRange;
+use Phpactor\LanguageServerProtocol\Range;
+
+class EvaluatableExpressionHandler implements Handler, CanRegisterCapabilities
+{
+    public function __construct(
+        private Workspace $workspace,
+        private Parser $parser,
+    ) {
+    }
+
+    public function methods(): array
+    {
+        return [
+            'textDocument/xevaluatableExpression' => 'xevaluatableExpression',
+        ];
+    }
+
+    /**
+     * @return Promise<EvaluatableExpression|null>
+     */
+    public function xevaluatableExpression(
+        TextDocumentIdentifier $textDocument,
+        Position $position
+    ): Promise {
+        return call(function () use ($textDocument, $position) {
+            $document = $this->workspace->get($textDocument->uri);
+            $offset = PositionConverter::positionToByteOffset($position, $document->text);
+            $document = TextDocumentBuilder::create($document->text)
+                ->uri($document->uri)
+                ->language('php')
+                ->build();
+
+            $char = substr($document, $offset->toInt(), 1);
+
+            // do not provide evaluatable for whitespace
+            if (trim($char) == '') {
+                return null;
+            }
+
+            $rootNode = $this->parser->parseSourceFile((string) $document, $document->uri()?->__toString());
+            $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
+            $ret = $this->nodeToEvaluatable($node);
+            return $ret;
+        });
+    }
+
+    public function registerCapabiltiies(ServerCapabilities $capabilities): void
+    {
+        $capabilities->experimental['xevaluatableExpressionProvider'] = true;
```

Comment:
I'm not sure why LSP defines this as `any` (i.e. `mixed`) is it guaranteed to be an array at this point? should we throw an exception or ignore it if it's not?

Reply:
It worked, but I do agree this could be a problem. It's defined as `any` so people can add in it whatever they want. TypeScript compiler/linter still [complains](https://github.com/phpactor/vscode-phpactor/pull/167/files#diff-04bba6a35cad1c794cbbe677678a51de13441b7a6ee8592b7b50be1f05c6f626R62) though. I need to investigate code patterns from Microsoft there...

What would be a good solution here?
```php
$capabilities->experimental ??= [];
$capabilities->experimental['xevaluatableExpressionProvider'] = true;
```

When the `InitializeMiddleware` calls the `registerCapabiltiies` method, `$capabilities->experimental` is currently `null`.

Reply:
hm. probably we should make an opinion and initiailize the `expreimental` field as an `array` ~or better a class with `public $xevluatableExpressionProvider`)~. The problem is that the `ServerCapabiltiies` class is instantiated in the Language Server `InitializeMiddleware`, so we can either:

- Take ownership of the `InitializeMiddleware` from the Phpactor Language Server package (i.e. copy and paste the class and change it)
- Introduce a factory for `ServerCapabilities` in the Language Server package.
- Take an opinion and unconditionally initialize it as an array in that class.

Alternatively we can do what you suggest but also throw an exception if it's neither `null` nor `array`, but it's not a real solution.

Reply:
On second thoughts, those are only partial solutions because `ServerCapabiltiies#experimental` will always be `mixed` statically. although you could more legitimately ignore the error then.

for now maybe just do what you suggest and either ignore the phpstan error or a type assertion.